dispatcher/dispatcher.go

```diff
@@ -0,0 +1,79 @@
+package dispatcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+var ErrInvalidValue = errors.New("invalid value")
+
+type Dispatcher interface {
+	Do(ctx context.Context, maxIter int, callback func()) error
+}
+
+type dispatcher struct {
+	numWorker int
+	sem       *semaphore.Weighted
+}
+
+// Do concurrently executes callback at most maxIter times or until ctx is done
+// or canceled. Concurrency is handled leveraging the semaphore pattern, which
+// ensures at most Dispatcher.numWorkers goroutines are spawned at the same time.
+func (d dispatcher) Do(ctx context.Context, maxIter int, callback func()) error {
+	if err := d.validate(maxIter, callback); err != nil {
+		return err
+	}
+
+	wg := sync.WaitGroup{}
+
+	for i := 0; i < maxIter || maxIter == 0; i++ {
+		wg.Add(1)
+
+		if err := d.sem.Acquire(ctx, 1); err != nil {
+			// err is either context.DeadlineExceeded or context.Canceled
+			// which are expected values so we stop the process silently.
+			wg.Done()
+			break
+		}
+
+		go func() {
+			defer func() {
+				d.sem.Release(1)
+				wg.Done()
+			}()
+			callback()
+		}()
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (d dispatcher) validate(maxIter int, callback func()) error {
+	if maxIter < 1 {
+		return fmt.Errorf("%w: maxIter: must be < 1, got %d", ErrInvalidValue, maxIter)
```

Comment:
`return fmt.Errorf("%w: maxIter: must be >= 1, got %d", ErrInvalidValue, maxIter)`

Reply:
Good catch! Not minor at all, totally forgot that use case. Fixed!